File: src/exoticatechnologies/util/FleetMemberUtils.java
package exoticatechnologies.util;

import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipVariantAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import exoticatechnologies.campaign.listeners.CampaignEventListener;
import org.apache.log4j.Logger;

import java.util.HashMap;
import java.util.Map;


public class FleetMemberUtils {
    public static final Map<String, FleetMemberAPI> moduleMap = new HashMap<>();
    private static final Logger log = Logger.getLogger(FleetMemberUtils.class);

    private FleetMemberUtils() {
    }

    public static FleetMemberAPI findMemberFromShip(ShipAPI ship) {
        String id = ship.getVariant().getHullVariantId();

        if(moduleMap.containsKey(id)) {
            return moduleMap.get(id);
        }

        if(ship.getParentStation() != null) {
            return findMemberFromShip(ship.getParentStation());
        }

        if(ship.getFleetMember() != null) {
            return ship.getFleetMember();
        }

        return findMemberForStats(ship.getMutableStats());
    }

    public static FleetMemberAPI findMemberForStats(MutableShipStatsAPI stats) {
        String id = stats.getVariant().getHullVariantId();

        if(moduleMap.containsKey(id)) {
            return moduleMap.get(id);
        }

        if (stats.getFleetMember() != null) {
            return stats.getFleetMember();
        }

        if (stats.getEntity() instanceof ShipAPI) {
            ShipAPI ship = (ShipAPI) stats.getEntity();
            if (ship.getFleetMember() != null) {
                return ship.getFleetMember();
            }
        }

        //note: this looks awful, but it actually doesn't go into this loop all that often.
        for (CampaignFleetAPI fleet : CampaignEventListener.Companion.getActiveFleets()) {
            FleetMemberAPI fm = searchFleetForStats(fleet, stats);
            if(fm != null) {
                return fm;
            }
        }

        return null;
    }

    private static FleetMemberAPI searchFleetForStats(CampaignFleetAPI fleet, MutableShipStatsAPI stats) {
        if(fleet == null) {
            return null;
        }

        for(FleetMemberAPI member : fleet.getFleetData().getMembersListCopy()) {
            if(member.isFighterWing()) continue;

            MutableShipStatsAPI memberStats = member.getStats();
            if(memberStats == stats) {
                return member;
            } else if (stats.getEntity() != null && memberStats.getEntity() == stats.getEntity()) {
                return member;
            } else if (stats.getFleetMember() != null && stats.getFleetMember() == member) {
                return member;
            } else if (stats.getVariant() != null && member.getVariant() == stats.getVariant()) {
                return member;
            } else {
                MutableShipStatsAPI opStats = null;
                try {
                    opStats = member.getVariant().getStatsForOpCosts();
                } catch (Throwable ex) {
                    // do nothing
                }

                if (opStats != null) {
                    if (opStats == stats) {
                        return member;
                    } else if (stats.getEntity() != null && opStats.getEntity() == stats.getEntity()) {
                        return member;
                    } else if (stats.getFleetMember() != null && opStats.getFleetMember() == stats.getFleetMember()) {
                        return member;
                    } else if (stats.getVariant() != null && opStats.getVariant() == stats.getVariant()) {
                        return member;
                    }
                }
            }

            ShipVariantAPI shipVariant = member.getVariant();
            for (String moduleVariantId : shipVariant.getStationModules().keySet()) {
                ShipVariantAPI moduleVariant = shipVariant.getModuleVariant(moduleVariantId);

                MutableShipStatsAPI moduleStats;
                try {
                    moduleStats = moduleVariant.getStatsForOpCosts();
                } catch (Throwable ex) {
                    continue;
                }
                if (moduleStats != null) {
                    if (moduleStats == stats) {
                        return member;
                    } else if (stats.getEntity() != null && stats.getEntity() == moduleStats.getEntity()) {
                        return member;
                    } else if (stats.getFleetMember() != null &&moduleStats.getFleetMember() == stats.getFleetMember()) {
                        return member;
                    }
                }
            }
        }

        return null;
    }
}

File: src/exoticatechnologies/util/FleetMemberUtils.kt
package exoticatechnologies.util

import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.characters.PersonAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipVariantAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import exoticatechnologies.campaign.listeners.CampaignEventListener.Companion.activeFleets
import exoticatechnologies.util.FleetMemberUtils.findFleetForVariant

object FleetMemberUtils {
    @JvmField
    val moduleMap: MutableMap<String, FleetMemberAPI> = HashMap()

    @JvmStatic
    fun findMemberFromShip(ship: ShipAPI): FleetMemberAPI? {
        val id = ship.variant.hullVariantId
        if (moduleMap.containsKey(id)) {
            return moduleMap[id]
        }
        if (ship.parentStation != null) {
            return findMemberFromShip(ship.parentStation)
        }
        return if (ship.fleetMember != null) {
            ship.fleetMember
        } else findMemberForStats(ship.mutableStats)
    }

    @JvmStatic
    fun findMemberForStats(stats: MutableShipStatsAPI): FleetMemberAPI? {
        val id = stats.variant.hullVariantId
        if (moduleMap.containsKey(id)) {
            return moduleMap[id]
        }
        if (stats.fleetMember != null) {
            return stats.fleetMember
        }
        if (stats.entity is ShipAPI) {
            val ship = stats.entity as ShipAPI
            if (ship.fleetMember != null) {
                return ship.fleetMember
            }
        }

        //note: this looks awful, but it actually doesn't go into this loop all that often.
        for (fleet in activeFleets) {
            val fm = searchFleetForStats(fleet, stats)
            if (fm != null) {
                return fm
            }
        }
        return null
    }

    private fun searchFleetForStats(fleet: CampaignFleetAPI?, stats: MutableShipStatsAPI): FleetMemberAPI? {
        if (fleet == null) {
            return null
        }
        for (member in fleet.fleetData.membersListCopy) {
            if (member.isFighterWing) continue
            val memberStats = member.stats
            if (memberStats === stats) {
                return member
            } else if (stats.entity != null && memberStats.entity === stats.entity) {
                return member
            } else if (stats.fleetMember != null && stats.fleetMember === member) {
                return member
            } else if (stats.variant != null && member.variant === stats.variant) {
                return member
            } else {
                var opStats: MutableShipStatsAPI? = null
                try {
                    opStats = member.variant.statsForOpCosts
                } catch (ex: Throwable) {
                    // do nothing
                }
                if (opStats != null) {
                    if (opStats === stats) {
                        return member
                    } else if (stats.entity != null && opStats.entity === stats.entity) {
                        return member
                    } else if (stats.fleetMember != null && opStats.fleetMember === stats.fleetMember) {
                        return member
                    } else if (stats.variant != null && opStats.variant === stats.variant) {
                        return member
                    }
                }
            }
            val shipVariant = member.variant
            for (moduleVariantId in shipVariant.stationModules.keys) {
                val moduleVariant = shipVariant.getModuleVariant(moduleVariantId)
                var moduleStats: MutableShipStatsAPI?
                moduleStats = try {
                    moduleVariant.statsForOpCosts
                } catch (ex: Throwable) {
                    continue
                }
                if (moduleStats != null) {
                    if (moduleStats === stats) {
                        return member
                    } else if (stats.entity != null && stats.entity === moduleStats.entity) {
                        return member
                    } else if (stats.fleetMember != null && moduleStats.fleetMember === stats.fleetMember) {
                        return member
                    }
                }
            }
        }
        return null
    }

    fun findFleetForVariant(variant: ShipVariantAPI, member: FleetMemberAPI): CampaignFleetAPI? {
        val id = variant.hullVariantId
        if (moduleMap.containsKey(id)) {
            return moduleMap[id]!!.fleetData.fleet
        }

        member.fleetData?.fleet?.let {
            return it
        }

        member.fleetCommander?.fleet?.let {
            return it
        }

        member.captain?.fleet?.let {
            return it
        }

        return activeFleets //do not remove the filterNotNull
            .filterNotNull()
            .firstOrNull { fleet ->
            fleet.membersWithFightersCopy?.any { member ->
                member.variant.stationModules.keys.any {
                    member.variant.getModuleVariant(it) == variant }
            } ?: false
        }
    }

    fun getFleetCommander(member: FleetMemberAPI): PersonAPI? {
        var fleetCommander = member.fleetCommander
        if (member.fleetCommander == null) {
            fleetCommander = if (member.fleetData != null) {
                member.fleetData.commander
            } else {
                findFleetForVariant(member.variant, member)?.commander
            }
        }

        return fleetCommander
    }

}

fun FleetMemberAPI.getFleetModuleSafe(): CampaignFleetAPI? {
    return findFleetForVariant(this.variant, this)
}